web/source/osk/activeLayout.ts

```diff
@@ -140,17 +207,154 @@ namespace com.keyman.osk {
+        totalMass += keyProbs[key] = 1 / (keyDists[key] * keyDists[key] + 1e-6); // Prevent div-by-0 errors.
+      }
+
+      for(let key in keyProbs) {
+        keyProbs[key] /= totalMass;
+      }
+
+      return keyProbs;
+    }
+
+    /**
+     * Computes a squared 'pseudo-distance' for the touch from each key.  (Not a proper metric.)
+     * Intended for use in generating a probability distribution over the keys based on the touch input.
+     * @param touchCoords A proportional (x, y) coordinate of the touch within the keyboard's geometry.
+     *                           Should be within [0, 0] to [1, 1].
+     * @param kbdScaleRatio The ratio of the keyboard's horizontal scale to its vertical scale.
+     *                           For a 400 x 200 keyboard, should be 2.
+     */
+    private keyTouchDistances(touchCoords: {x: number, y: number}, kbdScaleRatio: number): {[keyId: string]: number} {
+      let layer = this;
+
+      let keyDists: {[keyId: string]: number} = {};
+
+      // This double-nested loop computes a pseudo-distance for the touch from each key.  Quite useful for
+      // generating a probability distribution.
+      this.row.forEach(function(row: ActiveRow): void {
+        row.key.forEach(function(key: ActiveKey): void {
+          // These represent the within-key distance of the touch from the key's center.
+          // Both should be on the interval [0, 0.5].
+          let dx = Math.abs(touchCoords.x - key.proportionalX);
+          let dy = Math.abs(touchCoords.y - row.proportionalY);
+
+          // If the touch isn't within the key, these store the out-of-key distance
+          // from the closest point on the key being checked.
+          let distX: number, distY: number;
+
+          if(dx > 0.5 * key.proportionalWidth) {
+            distX = (dx - 0.5 * key.proportionalWidth);
+            dx = 0.5;
+          } else {
+            distX = 0;
+            dx /= key.proportionalWidth;
+          }
+
+          if(dy > 0.5 * layer.rowProportionalHeight) {
+            distY = (dy - 0.5 * layer.rowProportionalHeight);
+            dy = 0.5;
+          } else {
+            distY = 0;
+            dy /= layer.rowProportionalHeight;
+          }
+
+          // Now that the differentials are computed, it's time to do distance scaling.
+          //
+          // For out-of-key distance, we scale the X component by the keyboard's aspect ratio
+          // to get the actual out-of-key distance rather than proportional.
+          distX *= kbdScaleRatio;
+
+          // While the keys are rarely perfect squares, we map all within-key distance
+          // to a square shape.  (ALT/CMD should seem as close to SPACE as a 'B'.)
+          //
+          // For that square, we take the rowHeight as its edge lengths.
+          distX += dx * layer.rowProportionalHeight;
+          distY += dy * layer.rowProportionalHeight;
+
+          let distance = Math.sqrt(distX * distX + distY * distY);
```

Comment:
Later on, `keyDists[key` is used in the expression: 

```js
 totalMass += keyProbs[key] = 1 / (keyDists[key] * keyDists[key] + 1e-6); // Prevent div-by-0 errors.
```

It seems to me that squared euclidean distance is fine -- we can get rid of the `Math.sqrt()` since it doesn't really benefit us if we just square the value later anyway!

Reply:
Good catch - I was sure I'd already removed it.  Will do.